xprof/tally.cpp

```diff
@@ -32,9 +34,60 @@ void print_metadata(std::vector<std::string> metadata) {
     std::cout << value << std::endl;
 }
 
+void get_backend_id_from_name(const char *name, int *identifier){
+  *identifier = -1;
+  for(int i = 0; i < BACKEND_MAX; ++i)
+    // backend_name is located in xprof_utils.cpp
+    if (strcmp(backend_name[i],name) == 0) *identifier = i;
+}
+
+// Reference: https://favtutor.com/blogs/split-string-cpp
+std::vector <std::string> split(std::string str, char separator) {
+  std::vector <std::string> strings;
+  int startIndex = 0, endIndex = 0;
+  for (unsigned i = 0; i <= str.size(); i++) {
+    // If we reached the end of the word or the end of the input.
+    if (str[i] == separator || i == str.size()) {
+        endIndex = i;
+        std::string temp;
+        temp.append(str, startIndex, endIndex - startIndex);
+        strings.push_back(temp);
+        startIndex = endIndex + 1;
+    }
+  }
+
+  return strings;
+}
+
 void btx_initialize_usr_data(void *btx_handle, void **usr_data) {
   /* User allocates its own data structure */
   *usr_data = new tally_dispatch_t;
+  tally_dispatch_t *data = (tally_dispatch_t *)(*usr_data);
+
+  /* Backend information must match enum backend_e in xprof_utils.hpp */
+  data->backend_level = {
+    2, // BACKEND_UNKNOWN
+    2, // BACKEND_ZE
+    2, // BACKEND_OPENCL
+    2, // BACKEND_CUDA
+    1, // BACKEND_OMP_TARGET_OPERATIONS
+    0, // BACKEND_OMP
+    2  // BACKEND_HIP
+  };
+
+  // TODO: We must change this by data->params, but we need to adjust
+  // metababel to load btx_read_params before btx_initialize_usr_data.
+  std::vector<std::string> backends = split(std::string(((common_data_t *)btx_handle)->btx_params->backend_level), ',');
+
+  for (auto bk: backends) {
+    auto b_l =  split(bk, ':');
```

Comment:
please put an `assert(b_l.size()==2)`

And then you can do a b_l[0] and b_l[1]` if you prefer, front and back and kind of werid but whatever you prefer